feat: better lambda computation

nelimee opened the pull request:
## 🔗 Closed Issues

Towards #66

---

## 📝 Description

Improve the numerical methods used to compute lambda from the logical error probabilities per rounds obtained from the raw logical error probabilities.

---

## 🚦 Status

Ready to merge.

---

## 🛠️ Future Work

- Implement new methods to have a full pipeline to compute the error-budget for a given noise model.

---

## ➕️ Additional Information

This PR has already been reviewed by several people internally. More reviews are welcomed.

---

## 🧾 Release Note

PR title

mdhaber commented on deltakit-explorer/src/deltakit_explorer/analysis/_analysis.py:
```diff
@@ -514,99 +511,196 @@ def calculate_lep_and_lep_stddev(
+        )
+
+    logleppr = np.log(lep_per_round)
+    logleppr_stddev = lep_stddev_per_round / lep_per_round
```
Just to confirm, this is also making use of https://en.wikipedia.org/wiki/Propagation_of_uncertainty - if $f = \log{A} \rightarrow \sigma_f = \frac{\sigma_A}{A}$?

mdhaber commented on deltakit-explorer/src/deltakit_explorer/analysis/_analysis.py:
```diff
@@ -514,99 +511,196 @@ def calculate_lep_and_lep_stddev(
+    # Note that there are two ways to fit here:
+    # 1. Like what is done below, fit w.r.t the distance ``d``.
+    # 2. Fit w.r.t ``(d + 1) / 2``.
+    # Option 2 leads to simpler formulas, especially for standard deviation. But
```
```python3
import numpy as np
from scipy import optimize
from deltakit.explorer.analysis import calculate_lambda_and_lambda_stddev

rng = np.random.default_rng(821372777259345982)
lambda0 = 0.1
lambda_ = 2

d = np.asarray([5, 7, 9])

lepprs_stddev = 1e-3
lepprs = 1 / (lambda0 * lambda_ ** ((d + 1) / 2))
lepprs += rng.normal(0, lepprs_stddev, lepprs.shape)

lepprs_stddev = np.full_like(lepprs, lepprs_stddev)

def f(x, lambda_, lambda0):
    return 1 / (lambda0 * lambda_ ** ((d + 1) / 2))

res = calculate_lambda_and_lambda_stddev(d, lepprs, lepprs_stddev)
p, cov = optimize.curve_fit(f, d, lepprs, sigma=lepprs_stddev, absolute_sigma=True)

print(res.lambda_, p[0])  # 1.9992504908511501 1.9992642173284623
print(res.lambda0, p[1])  # 0.10006000676191887 0.10005792584295113
print(res.lambda_stddev, np.sqrt(cov[0, 0]))  # 0.0025464304928163348 0.0025492573171015005
print(res.lambda0_stddev, np.sqrt(cov[1, 1]))  # 0.00036742471543639804 0.0004250067464587885
```

As we might expect, there is not perfect agreement in the presence of noise because minimizing the $L^2$ norm is going to be different in the original space vs log-space. When i turn off the noise, there is very close agreement between `lambda_`, `lambda0`, and `lambda_stddev`, but `lambda0_stddev` is still a bit different. I'm not sure if that should be expected?

nelimee commented on deltakit-explorer/src/deltakit_explorer/analysis/_analysis.py:
```diff
@@ -514,99 +511,196 @@ def calculate_lep_and_lep_stddev(
 
+@dataclass(frozen=True)
+class LambdaResults:
+    """Named-tuple-like class containing computation results from
```
Will work on deprecation today, hopefully that will be a small PR that can be merged before this one.

nelimee commented on deltakit-explorer/src/deltakit_explorer/analysis/_analysis.py:
```diff
@@ -514,99 +511,192 @@ def calculate_lep_and_lep_stddev(
+    # Option 2 leads to simpler formulas, especially for standard deviation. But
+    # numerical investigations have found that option 2 was behaving very poorly
+    # (several orders of magnitude worse than option 1) when Λ0 is close to ``1``. For
+    # that reason, option 1 is preferred below.
```
Yep, that's also what I am witnessing at the moment. This behaviour was only apparent for $\Lambda_0 = 1$ (or **very** close).

The plot I had: 
<img width="556" height="802" alt="Screenshot 2025-10-15 at 11 46 20" src="https://github.com/user-attachments/assets/06120603-5c1d-44d7-b22a-7e0e1da3b586" />
With:
- X-axis is the value of $\Lambda$
- Y-axis is the absolute error between the ground truth value of $\Lambda$ (i.e., the X-axis value) and the computed value for `d`-fit (orange) and `(d+1)/2`-fit (blue).

See my upcoming message on this PR.

nelimee commented:
So I have played with the different fitting methods to see if there was a clear winner. Code is messy and slightly too long for this PR, so hosted in [this gist](https://gist.github.com/nelimee/21a4f08ba0df44465df2c7eab90801e9).

I made a plot with `num_points = 300` because that was done in a reasonable time (10 minutes timescale).
<img width="1920" height="985" alt="lambda_fitting_methods" src="https://github.com/user-attachments/assets/12c78485-7ba1-4926-9f1f-bc504165e11b" />

Explanation of the plot:
- X-axis is the exact value of $\Lambda$ used to generate noisy logical error probability per round data to fit.
- Y-axis is the exact value of $\Lambda_0$.
- Noise has been added by sampling from `N(0, 0.1)` and multiplying the synthetic logical error probability per round data (obtained with the exact value of $\Lambda$ and $\Lambda_0$) by `1 + err` to make it a relative stddev.

For rows:
1. $\Lambda$ values
2. $\Lambda_0$ values
3. $\Lambda$ standard deviations
4. $\Lambda_0$ standard deviations

The first column show, for each point, the method that obtained the best result (closest to the ground truth value for $\Lambda$ and $\Lambda_0$, closest to `0` for standard-deviations).

For the two top rows (about $\Lambda$ and $\Lambda_0$), the 2nd (resp, 3rd and 4th) columns show the absolute error between the true value and the estimated value for the method `d` (resp, `(d+1)/2`, `direct`).

For the two top rows (about $\Lambda$ and $\Lambda_0$), the 2nd (resp, 3rd and 4th) columns show the relative error between the standard deviation of method `d` and the estimated standard deviation for the method `d` (resp, `(d+1)/2`, `direct`).

White points means `0` (and do not appear because colormap is in log-scale, so `0` is `NaN`).

Conclusions:
- For $\Lambda$ value, all 3 methods are acceptable and give very accurate results.
- For $\Lambda_0$ value, all 3 methods are acceptable and give very accurate results.
- For $\Lambda$ standard deviation, the `(d+1)/2` method seems to give the lowest value.
- For $\Lambda_0$ standard deviation, all 3 methods give comparable standard deviation.

I think that shows that there is no clear winner. Hope I did not let a mistake slip in the code.

PS: the 3 methods pass the unit-tests in `test_analysis.py`.

nelimee commented:
The comment from @mdhaber about adding covariance in the computation has been implemented and all 3 methods now show remarkable agreement in both estimated values and standard deviations.

nelimee commented:
Since the last approval, I have:

- fixed small parts of the documentation on lambda computation.
- added missing documentation of parameters in LEPPR computation (forgot to do it in #126).
- added the `@deprecated` decorator to `get_exp_fit` and `get_lambda_fit` that can now be replaced by the newly introduced functions. Tests for these functions, tests now expect a `DeprecationWarning` to be raised. I marked these functions as being removed no earlier than `deltakit` version `1.0.0`.

These are relatively minor, so I will merge this PR once CI pass.
